File: checkout_system.py
#!/usr/bin/env python3
"""
Multi-Client Checkout System
A store checkout system that handles multiple customers sequentially.
"""

# Store products with fixed prices
PRODUCTS = {
    1: {"name": "Keyboard", "price": 100.00},
    2: {"name": "Mouse", "price": 50.00},
    3: {"name": "Monitor", "price": 300.00},
    4: {"name": "USB Cable", "price": 20.00},
    5: {"name": "Headphones", "price": 150.00}
}

TAX_RATE = 0.13  # 13% tax rate


def show_menu():
    """Display the store menu with available products."""
    print("Welcome to the store!")
    print("Available products:")
    for product_id, product_info in PRODUCTS.items():
        print(f"{product_id}. {product_info['name']} - ${product_info['price']:.0f}")
    print()


def calculate_tax(subtotal):
    """Calculate tax based on subtotal."""
    return subtotal * TAX_RATE


def calculate_discount(subtotal):
    """Calculate discount based on subtotal amount."""
    # Based on the example, no discount is applied for this scenario
    # You can modify this logic as needed for your business rules
    if subtotal >= 1000:
        return subtotal * 0.10  # 10% discount for orders $1000+
    else:
        return 0.00  # No discount for orders under $1000


def calculate_delivery_fee(distance):
    """Calculate delivery fee based on distance."""
    if distance <= 10:
        return 15.00  # Base fee for up to 10km
    elif distance <= 20:
        return 30.00  # Fee for 11-20km
    else:
        return 45.00  # Fee for 21km+


def get_valid_product_choice():
    """Get a valid product choice from the user."""
    while True:
        try:
            choice = int(input("Enter product number to add to cart (0 to finish): "))
            if choice == 0:
                return 0
            elif choice in PRODUCTS:
                return choice
            else:
                print(f"Invalid product number. Please choose from 1-{len(PRODUCTS)} or 0 to finish.")
        except ValueError:
            print("Please enter a valid number.")


def get_valid_quantity():
    """Get a valid quantity from the user."""
    while True:
        try:
            quantity = int(input("Enter quantity: "))
            if quantity > 0:
                return quantity
            else:
                print("Quantity must be greater than 0.")
        except ValueError:
            print("Please enter a valid number.")


def get_valid_distance():
    """Get a valid delivery distance from the user."""
    while True:
        try:
            distance = float(input("Enter delivery distance in km: "))
            if distance > 0:
                return distance
            else:
                print("Distance must be greater than 0.")
        except ValueError:
            print("Please enter a valid number.")


def get_delivery_choice():
    """Get whether customer wants delivery."""
    while True:
        choice = input("Do you want delivery? (yes/no): ").lower().strip()
        if choice in ['yes', 'y']:
            return True
        elif choice in ['no', 'n']:
            return False
        else:
            print("Please enter 'yes' or 'no'.")


def process_customer():
    """Process a single customer's order."""
    # Show menu
    show_menu()
    
    # Shopping cart to store items
    cart = {}
    
    # Shopping loop
    while True:
        product_choice = get_valid_product_choice()
        
        if product_choice == 0:
            break
            
        quantity = get_valid_quantity()
        
        # Add to cart
        if product_choice in cart:
            cart[product_choice] += quantity
        else:
            cart[product_choice] = quantity
        
        print()  # Add blank line for readability
    
    # If cart is empty, skip to next customer
    if not cart:
        print("No items in cart. Moving to next customer...\n")
        return
    
    # Get customer information
    print()
    name = input("Please enter your name: ").strip()
    phone = input("Please enter your phone number: ").strip()
    print()
    
    # Calculate subtotal
    subtotal = 0
    for product_id, quantity in cart.items():
        subtotal += PRODUCTS[product_id]["price"] * quantity
    
    # Calculate tax and discount
    tax = calculate_tax(subtotal)
    discount = calculate_discount(subtotal)
    
    # Display order summary
    print(f"Subtotal: ${subtotal:.2f}")
    print(f"Tax ({TAX_RATE*100:.0f}%): ${tax:.2f}")
    print(f"Discount: ${discount:.2f}")
    print()
    
    # Handle delivery
    delivery_fee = 0
    if get_delivery_choice():
        distance = get_valid_distance()
        delivery_fee = calculate_delivery_fee(distance)
        print(f"Delivery fee: ${delivery_fee:.2f}")
    
    # Calculate total
    total = subtotal + tax - discount + delivery_fee
    
    # Display final receipt
    print("\n" + "-" * 28)
    print(f"Total amount to pay: ${total:.2f}")
    print()
    print(f"Thank you for shopping with us, {name}!")
    print("-" * 41)
    print()


def main():
    """Main program loop to handle multiple customers."""
    print("🛒 Multi-Client Checkout System")
    print("=" * 35)
    print()
    
    try:
        while True:
            process_customer()
            print("Ready for next customer...")
            print()
            
            # Optional: Ask if there's another customer
            # Uncomment the following lines if you want to add an exit option
            # continue_choice = input("Is there another customer? (yes/no): ").lower().strip()
            # if continue_choice in ['no', 'n']:
            #     break
            # print()
            
    except KeyboardInterrupt:
        print("\n\nThank you for using our checkout system!")
        print("Have a great day!")


if __name__ == "__main__":
    main()

File: checkout_system.dart
import 'dart:io';

// Individual product variables - Product 1 (Keyboard)
int product1Id = 1;
String product1Name = 'Keyboard';
double product1Price = 100.0;

// Individual product variables - Product 2 (Mouse)
int product2Id = 2;
String product2Name = 'Mouse';
double product2Price = 50.0;

// Individual product variables - Product 3 (Monitor)
int product3Id = 3;
String product3Name = 'Monitor';
double product3Price = 300.0;

// Individual product variables - Product 4 (USB Cable)
int product4Id = 4;
String product4Name = 'USB Cable';
double product4Price = 20.0;

// Individual product variables - Product 5 (Headphones)
int product5Id = 5;
String product5Name = 'Headphones';
double product5Price = 150.0;

// Cart variables - individual quantities for each product
int cartQuantity1 = 0; // Keyboard quantity
int cartQuantity2 = 0; // Mouse quantity
int cartQuantity3 = 0; // Monitor quantity
int cartQuantity4 = 0; // USB Cable quantity
int cartQuantity5 = 0; // Headphones quantity

// Customer info variables
String customerName = '';
String customerPhone = '';

void showMenu() {
  print('\nWelcome to the store!');
  print('Available products:');
  print('$product1Id. $product1Name - \$${product1Price.toStringAsFixed(0)}');
  print('$product2Id. $product2Name - \$${product2Price.toStringAsFixed(0)}');
  print('$product3Id. $product3Name - \$${product3Price.toStringAsFixed(0)}');
  print('$product4Id. $product4Name - \$${product4Price.toStringAsFixed(0)}');
  print('$product5Id. $product5Name - \$${product5Price.toStringAsFixed(0)}');
}

void addToCart() {
  while (true) {
    stdout.write('\nEnter product number to add to cart (0 to finish): ');
    String? input = stdin.readLineSync();
    
    if (input == null || input.isEmpty) {
      print('Invalid input. Please try again.');
      continue;
    }
    
    int productNumber;
    try {
      productNumber = int.parse(input);
    } catch (e) {
      print('Invalid input. Please enter a number.');
      continue;
    }
    
    if (productNumber == 0) {
      break;
    }
    
    if (productNumber < 1 || productNumber > 5) {
      print('Invalid product number. Please choose from 1-5.');
      continue;
    }
    
    stdout.write('Enter quantity: ');
    String? quantityInput = stdin.readLineSync();
    
    if (quantityInput == null || quantityInput.isEmpty) {
      print('Invalid quantity. Please try again.');
      continue;
    }
    
    int quantity;
    try {
      quantity = int.parse(quantityInput);
    } catch (e) {
      print('Invalid quantity. Please enter a number.');
      continue;
    }
    
    if (quantity <= 0) {
      print('Quantity must be greater than 0.');
      continue;
    }
    
    // Add to cart based on product number
    String productName = '';
    if (productNumber == 1) {
      cartQuantity1 += quantity;
      productName = product1Name;
    } else if (productNumber == 2) {
      cartQuantity2 += quantity;
      productName = product2Name;
    } else if (productNumber == 3) {
      cartQuantity3 += quantity;
      productName = product3Name;
    } else if (productNumber == 4) {
      cartQuantity4 += quantity;
      productName = product4Name;
    } else if (productNumber == 5) {
      cartQuantity5 += quantity;
      productName = product5Name;
    }
    
    print('Added ${quantity}x $productName to cart.');
  }
}

void getCustomerInfo() {
  stdout.write('\nPlease enter your name: ');
  customerName = stdin.readLineSync() ?? '';
  
  stdout.write('Please enter your phone number: ');
  customerPhone = stdin.readLineSync() ?? '';
}

double calculateSubtotal() {
  double total = 0.0;
  
  total += product1Price * cartQuantity1;
  total += product2Price * cartQuantity2;
  total += product3Price * cartQuantity3;
  total += product4Price * cartQuantity4;
  total += product5Price * cartQuantity5;
  
  return total;
}

double calculateTax(double subtotal) {
  return subtotal * 0.13; // 13% tax
}

double calculateDiscount(double subtotal) {
  // Discount rules: 10% for orders over $500, 5% for orders over $200
  if (subtotal > 500) {
    return subtotal * 0.10;
  } else if (subtotal > 200) {
    return subtotal * 0.05;
  }
  return 0.0;
}

double calculateDeliveryFee(double distance) {
  // Base fee $5 + $2 per km over 5km
  if (distance <= 5) {
    return 5.0;
  } else {
    return 5.0 + (distance - 5) * 2.0;
  }
}

bool getDeliveryChoice() {
  while (true) {
    stdout.write('\nDo you want delivery? (yes/no): ');
    String? input = stdin.readLineSync()?.toLowerCase();
    
    if (input == 'yes' || input == 'y') {
      return true;
    } else if (input == 'no' || input == 'n') {
      return false;
    } else {
      print('Please enter yes or no.');
    }
  }
}

double getDeliveryDistance() {
  while (true) {
    stdout.write('Enter delivery distance in km: ');
    String? input = stdin.readLineSync();
    
    if (input == null || input.isEmpty) {
      print('Invalid input. Please try again.');
      continue;
    }
    
    try {
      double distance = double.parse(input);
      if (distance <= 0) {
        print('Distance must be greater than 0.');
        continue;
      }
      return distance;
    } catch (e) {
      print('Invalid distance. Please enter a number.');
    }
  }
}

bool isCartEmpty() {
  return cartQuantity1 == 0 && cartQuantity2 == 0 && cartQuantity3 == 0 && 
         cartQuantity4 == 0 && cartQuantity5 == 0;
}

void printCartSummary() {
  print('\nYour cart:');
  if (cartQuantity1 > 0) {
    print('${cartQuantity1}x $product1Name - \$${(product1Price * cartQuantity1).toStringAsFixed(2)}');
  }
  if (cartQuantity2 > 0) {
    print('${cartQuantity2}x $product2Name - \$${(product2Price * cartQuantity2).toStringAsFixed(2)}');
  }
  if (cartQuantity3 > 0) {
    print('${cartQuantity3}x $product3Name - \$${(product3Price * cartQuantity3).toStringAsFixed(2)}');
  }
  if (cartQuantity4 > 0) {
    print('${cartQuantity4}x $product4Name - \$${(product4Price * cartQuantity4).toStringAsFixed(2)}');
  }
  if (cartQuantity5 > 0) {
    print('${cartQuantity5}x $product5Name - \$${(product5Price * cartQuantity5).toStringAsFixed(2)}');
  }
}

void processCheckout() {
  if (isCartEmpty()) {
    print('Your cart is empty. Please add some products first.');
    return;
  }
  
  printCartSummary();
  
  getCustomerInfo();
  
  double subtotal = calculateSubtotal();
  double tax = calculateTax(subtotal);
  double discount = calculateDiscount(subtotal);
  double deliveryFee = 0.0;
  
  print('\nSubtotal: \$${subtotal.toStringAsFixed(2)}');
  print('Tax (13%): \$${tax.toStringAsFixed(2)}');
  print('Discount: \$${discount.toStringAsFixed(2)}');
  
  bool wantsDelivery = getDeliveryChoice();
  if (wantsDelivery) {
    double distance = getDeliveryDistance();
    deliveryFee = calculateDeliveryFee(distance);
    print('Delivery fee: \$${deliveryFee.toStringAsFixed(2)}');
  }
  
  double total = subtotal + tax - discount + deliveryFee;
  
  print('\n----------------------------');
  print('Total amount to pay: \$${total.toStringAsFixed(2)}');
  print('\nThank you for shopping with us, $customerName!');
  print('-----------------------------------------');
}

void resetForNextCustomer() {
  cartQuantity1 = 0;
  cartQuantity2 = 0;
  cartQuantity3 = 0;
  cartQuantity4 = 0;
  cartQuantity5 = 0;
  customerName = '';
  customerPhone = '';
}

void runCheckoutSystem() {
  print('🛒 Multi-Client Checkout System');
  print('===============================');
  
  while (true) {
    showMenu();
    addToCart();
    
    if (!isCartEmpty()) {
      processCheckout();
    }
    
    print('\nReady for next customer...');
    print('Press Enter to continue or type "quit" to exit...');
    
    String? input = stdin.readLineSync()?.toLowerCase();
    if (input == 'quit' || input == 'q') {
      print('Thank you for using our checkout system!');
      break;
    }
    
    resetForNextCustomer();
  }
}

void main() {
  runCheckoutSystem();
}